File: app/src/main/java/applab/veiligthuis/repository/MeldingRepositoryImpl.java
package applab.veiligthuis.repository;

import static android.content.ContentValues.TAG;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import applab.veiligthuis.model.Melding;

public class MeldingRepositoryImpl {
    private DatabaseReference meldingenRef;
    private MutableLiveData<String> mSuccessMessage = new MutableLiveData<>();
    private MutableLiveData<String> mErrorMessage = new MutableLiveData<>();

    FirebaseDatabase firebaseDatabase;

    public LiveData<String> getSuccessMessage() {
        return mSuccessMessage;
    }

    public LiveData<String> getErrorMessage() {
        return mErrorMessage;
    }

    public MeldingRepositoryImpl(){
        firebaseDatabase = FirebaseDatabase.getInstance();
        meldingenRef = firebaseDatabase.getReference("Meldingen/");
        meldingenRef.keepSynced(true);
    }

    public void addMelding(Melding melding){

        String key = meldingenRef.push().getKey();
        melding.setKey(key);
        meldingenRef.child(key).setValue(melding)
                .addOnSuccessListener(aVoid -> mSuccessMessage.setValue("Melding opgeslagen in database."))
                .addOnFailureListener(e -> mErrorMessage.setValue("Fout bij opslaan melding: " + e.getMessage()));

    }

    public LiveData<List<Melding>> getMeldingenListLiveData(){
        MutableLiveData<List<Melding>> meldingenLiveData = new MutableLiveData<>();
        meldingenRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                List<Melding> meldingen = new ArrayList<>();
                for(DataSnapshot snap : snapshot.getChildren()){
                    Melding melding = snap.getValue(Melding.class);
                    if(melding != null){
                        meldingen.add(melding);
                    }
                }
                meldingenLiveData.setValue(meldingen);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.d(TAG, "onCancelled: ");
            }
        });
        return meldingenLiveData;
    }
}

File: app/src/main/java/applab/veiligthuis/repository/melding/MeldingRepositoryImpl.kt
package applab.veiligthuis.repository.melding


import android.util.Log
import applab.veiligthuis.domain.model.melding.Melding

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.ktx.database
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow

class MeldingRepositoryImpl : MeldingRepository {
    val database = Firebase.database
    val ref = database.getReference("Tests/meldingen/")

    init {
        ref.keepSynced(true)
    }

    override fun <T: Melding> getMeldingen(paths: List<String>, meldingtype: Class<T>): Flow<List<Melding?>> {
        return callbackFlow {
            val postListener = object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    val items = arrayListOf<Melding?>()
                    paths.forEach { path ->
                        snapshot.child(path).children.map { ds ->
                            val melding = ds.getValue(meldingtype)
                            items.add(melding)
                        }
                    }
                    trySend(items)
                }
                override fun onCancelled(error: DatabaseError) {
                    cancel()
                }
            }
            ref.addValueEventListener(postListener)
            awaitClose { ref.removeEventListener(postListener) }
        }
    }

    override fun <T : Melding> getMelding(meldingKey: String, meldingType: Class<T>): Flow<Melding> {
        return callbackFlow {
            val meldingListener = object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    val melding = snapshot.child(meldingKey).getValue(meldingType)
                    if (melding != null) {
                        trySend(melding)
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                    cancel()
                }
            }
            ref.addValueEventListener(meldingListener)
            awaitClose { ref.removeEventListener(meldingListener) }
        }
    }


    override fun addMelding(melding: Melding) {
        if(melding.key == null) {
            val key = ref.child(melding.getPaths()[0]).push().key
            if(key == null) {
                Log.w("REP", "Melding niet naar db gepushed.")
            }
            melding.key = key
        }
        editMelding(melding)
    }

    override fun editMelding(melding: Melding) {
        val childUpdates = hashMapOf<String, Any>()
        melding.getPaths().forEach { path ->
            childUpdates[path] = melding.toMap()
        }
        ref.updateChildren(childUpdates)
    }

    override fun deleteMelding(melding: Melding) {
        val childUpdates = hashMapOf<String, Any?>()
        melding.getPaths().forEach { path ->
            childUpdates[path] = null
        }
        ref.updateChildren(childUpdates)
    }

}

